### PR TITLE
Enforce resource limits for Soroban

### PR DESCRIPTION
**Pull request by dmkozh**
# Description

- Perform soroban transaction validation and ensure per-transaction limits aren't exceeded
- Plumb Soroban resources and network configuration to the host and enforce the resource limits at apply time
- Also contains supporting refactoring and minor fixes. The plumbing solution is not super elegant (pass `Application` to more functions), but is better

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/transactions/InvokeHostFunctionOpFrame.cpp [hunk 1]
```diff
@@ -321,6 +351,7 @@ InvokeHostFunctionOpFrame::doApply(AbstractLedgerTxn& ltx, Config const& cfg,
 
         keys.emplace(std::move(lk));
     }
+    metadataSizeBytes += metrics.mLedgerWriteByte;
```
Comment (sisuresh) [inline src/transactions/InvokeHostFunctionOpFrame.cpp:354]: It looks like you only validate `metadataSizeBytes` again if the call emitted events. If not, then `mLedgerWriteByte` could cause you to exceed the limit.

File: src/ledger/LedgerManagerImpl.cpp [hunk 2]
```diff
@@ -1093,19 +1093,24 @@ LedgerManagerImpl::maybeUpdateNetworkConfig(bool upgradeHappened,
...
-    if (!mLastSorobanNetworkConfig)
+    if (!mSorobanNetworkConfig)
     {
-        mLastSorobanNetworkConfig = std::make_optional<SorobanNetworkConfig>();
+        mSorobanNetworkConfig = std::make_optional<SorobanNetworkConfig>();
     }
-    LedgerTxn ltx(rootLtx, false, TransactionMode::READ_ONLY_WITHOUT_SQL_TXN);
-    if (protocolVersionStartsFrom(ltx.loadHeader().current().ledgerVersion,
-                                  SOROBAN_PROTOCOL_VERSION))
+    uint32_t ledgerVersion{};
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply src/ledger/LedgerManagerImpl.cpp:1104]: This is just to make sure we don't get uninitialized value warnings (and uninitialized value itself)

File: src/transactions/InvokeHostFunctionOpFrame.cpp (hunk 1 above)
Comment (dmkozh) [reply src/transactions/InvokeHostFunctionOpFrame.cpp:354]: Nice catch, fixed.

File: src/transactions/TransactionFrame.cpp [hunk 3]
```diff
@@ -523,6 +523,38 @@ TransactionFrame::validateSorobanOpsConsistency() const
...
+    if (resources.writeBytes > config.txMaxWriteBytes())
+    {
+        return false;
+    }
+    if (resources.extendedMetaDataSizeBytes >
+        config.txMaxExtendedMetaDataSizeBytes())
+    {
+        return false;
+    }
+    if (readEntries + writeEntries > config.txMaxReadLedgerEntries() ||
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply src/transactions/TransactionFrame.cpp:551]: I'd hope we reject such huge envelopes in overlay/tx queue. Also `size_t` should be 64-bit, while XDR size limit is 32-bit.

File: src/transactions/test/TxEnvelopeTests.cpp [hunk 4]
```diff
@@ -2465,3 +2466,112 @@ TEST_CASE_VERSIONS("txenvelope", "[tx][envelope]")
         for_versions_from(13, *app, [&] { doChecks(txSUCCESS); });
     }
 }
+
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+TEST_CASE("soroban txs not allowed before protocol upgrade",
+          "[tx][envelope][soroban]")
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply src/transactions/test/TxEnvelopeTests.cpp:2472]: I'm trying to switch 'contract' to 'soroban' where I notice it and I actually did update these tests to use the '[soroban]' tag.

[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/test/TxEnvelopeTests.cpp:2472]: Ah sorry I missed that

**Comment by sisuresh** [comment]
r+ 8f0d4614eb0f3ebd446689e813d68b14fa46a0cc

**Pull request merged**